### ISSUE TITLE
[CoE Starter Kit - Feature]: Admin | Sync Template v3 (Connection Identities) - PowerShell Alternative

### ISSUE THREAD

**Issue by Grant-Archibald-MS** [microsoft/coe-starter-kit#7906]
### Is your feature request related to a problem? Please describe.

As noted in [[CoE Starter Kit - BUG] Admin | Sync Template v3 (Connection Identities) The action 'Get_Connections_as_Admin' has an aggregated page results size of more than '209809819' bytes](https://github.com/microsoft/coe-starter-kit/issues/6276) there is a product limitation in retrieving connections using the [Power Apps for Admins](https://learn.microsoft.comconnectors/powerappsforadmins/)

### Describe the solution you'd like

Provide an example of how the cloud flow could be disabled and a PowerShell script using [Get-AdminPowerAppConnection](https://learn.microsoft.compowershell/module/microsoft.powerapps.administration.powershell/get-adminpowerappconnection?view=pa-ps-latest) to get all the connections in an environment

Extend the example to use PowerShell with Dataverse API [Use PowerShell and Visual Studio Code with the Dataverse Web API](https://learn.microsoft.com/power-apps/developer/data-platform/webapi/use-ps-and-vscode-web-api) or [Use bulk operation messages](https://learn.microsoft.com/power-apps/developer/data-platform/bulk-operations?tabs=sdk#upsertmultiple) from a simple C# command line application to upload the results from PowerShell to the [Connection Reference Identities](https://learn.microsoft.com/power-platform/guidance/coe/core-components#tables)

**Comment by Grant-Archibald-MS** [comment]
Suggested approach:
- Validate PowerShell returns information similar to the connector
- Create example of how to upload those results in batches to the Dataverse Table
- Replicate the issue by creating large number of connections in an environment
- Verify that the PowerShell command continues to work with large sizes above the 21MB issue above

**Comment by Jenefer-Monroe** [comment]
Here are some starters on how to use powershell to [find](https://learn.microsoft.com/powershell/module/microsoft.powerapps.administration.powershell/get-adminpowerappconnection?view=pa-ps-latest) and [delete](https://learn.microsoft.com/powershell/module/microsoft.powerapps.administration.powershell/remove-adminpowerappconnection?view=pa-ps-latest) unwanted connections. 

It will look something like this:
Get-AdminPowerAppConnection -EnvironmentName [Guid]

Remove-AdminPowerAppConnection -ConnectionName [ConnectionName] -ConnectorName shared_twitter -EnvironmentName Default-[Guid]

**Comment by TomekPi** [comment]
Is there any update on the above?

The PowerShell alternative is working for us, we can get connections, however we'd like to have it in CoE Dataverse.

Are there any plans to create some more automated workaround for this issue or are we on our own? Or maybe are there some rumors regarding V2 actions, similar to the "Get Flows" one?

**Comment by DaveTheTeamsGuy** [comment]
We have this problem and are also noticing that there are missing apps in our toolkit (visible in the MS admin console, but not anywhere in CoE).  Is it possible this issue is preventing app inventory, or would it be unrelated?

**Comment by Grant-Archibald-MS** [comment]
Also linking to #8031 [[CoE Starter Kit - Feature]: One off solution for too-many-connections](https://github.com/microsoft/coe-starter-kit/issues/8031)

**Comment by Grant-Archibald-MS** [comment]
@DaveTheTeamsGuy @TomekPi it would be good to geed feedback on #8031 and the importance for you to help build on the PowerShell approach that @Jenefer-Monroe discusses above

**Comment by DaveTheTeamsGuy** [comment]
@Grant-Archibald-MS I've done so, however my question while related to this issue is centered more around downstream results of this flow erroring out.

**Comment by Jenefer-Monroe** [comment]
Hello. Sorry lots of threads here. Can you clarify what you are asking? Which flow are you asking about?

**Comment by DaveTheTeamsGuy** [comment]
We have the "Get_Connections_as_Admin' has an aggregated page results size of more than '209809819' bytes" issue with the Admin | Sync Template v3 (Connection Identities) flow in two of our environments.  We also noticed that a number of apps are missing from our CoE toolkit apps inventory in these two environments.  Is it possible this issue is preventing those apps from being properly inventoried?

**Comment by DaveTheTeamsGuy** [comment]
Ugh you can probably disregard  - just found that Admin | Sync Template v4 (Apps) was not turned on.  PEBCAK...

**Comment by Jenefer-Monroe** [comment]
We do inventory apps / flows even if a connector for them cannot be found. 
For Connection Identities it means that you will not be able to gather the connection identitities in that envt. 

What a connection identity is: we knew that we did not want to store all the connections in the tenant in the CoE. For this same reason, there are so many that it would quickly overwhelm the kit and flood the dataverse. So instead we just store a "connection identity" 
the Connection Identity: Jon - Default - O365 Users
might actually map to 3 connections: 

1. Jon - Default - O365 - Errored
2. Jon - Default - O365 - Expirng soon
3. Jon - Default - O365 Connected

So the downstream impact of that you wont see them for cleanup in the new Managed Permissions app
![image](https://github.com/microsoft/coe-starter-kit/assets/89584709/e36727d5-89ef-4220-99c5-53d7b8548e9c)
![image](https://github.com/microsoft/coe-starter-kit/assets/89584709/fdebc976-a317-4f9e-85bd-b6c048e9b298)

**Comment by DaveTheTeamsGuy** [comment]
I agree, it seems to be a good workaround, we're looking into it.  I did try doing the Get- last Friday on one of our two affected environments and all it did was gradually increase the memory size of the Poweshell process to the point where I had to kill the process.  It never gave me a result.  We must really have a lot of bad connections in those environments.  Doing the Get- on an unaffected environment returned results, so there's something more going on there.

**Comment by Jenefer-Monroe** [comment]
oh boy. perhaps if you filter down to a specific connector that will return them and you can get cleaning that way. 
For example there are probably a lot of old o365 outlook/user/group type ones and those show with their error states correctly. So you could clean those.

**Comment by bilodeauju** [comment]
@Jenefer-Monroe thank you for clarifying the impact. My tenant already has 4 environments affected by this connector product limitation and so I was hoping that meanwhile I could use the new Managed Permissions app tool to preventively cleanup orphaned connections in other environments not yet affected. 

I tried to see if I could delete all orphaned connections in a specific environment at once using the Managed Permissions app, by filtering for one specific environment and then selecting all orphaned connection references for that environment, but there is no option to manage connection or delete connection when selecting multiple connection references. Does this tool only allow cleaning one orphaned connection at a time? If so that would not be feasible at all for my tenant has we have hundreds of orphaned connection per environments. Is there any easy way to quickly delete multiple orphaned connection at once that I am missing, or should I wait for https://github.com/microsoft/coe-starter-kit/issues/8031?

View when multiple connection references are selected:
![image](https://github.com/microsoft/coe-starter-kit/assets/81716086/daa9ba4d-34b5-41ce-b0be-3def4e6cd272)

View when only one connection reference is selected:
![image](https://github.com/microsoft/coe-starter-kit/assets/81716086/1701eb82-02d3-4ba2-9e0b-a4d33cf22f88)
![image](https://github.com/microsoft/coe-starter-kit/assets/81716086/0ff43670-1078-49db-9d08-5df7913425b4)


> We do inventory apps / flows even if a connector for them cannot be found. For Connection Identities it means that you will not be able to gather the connection identitities in that envt.
> 
> What a connection identity is: we knew that we did not want to store all the connections in the tenant in the CoE. For this same reason, there are so many that it would quickly overwhelm the kit and flood the dataverse. So instead we just store a "connection identity" the Connection Identity: Jon - Default - O365 Users might actually map to 3 connections:
> 
> 1. Jon - Default - O365 - Errored
> 2. Jon - Default - O365 - Expirng soon
> 3. Jon - Default - O365 Connected
> 
> So the downstream impact of that you wont see them for cleanup in the new Managed Permissions app

**Comment by Jenefer-Monroe** [comment]
That app will not be useable for environments impacted by this product limitation. You need to do the cleanup first.

**Comment by TomekPi** [comment]
@Jenefer-Monroe , regarding cleaning broken connections, I've came across #3761 where user mentioned that removing errored connections broke some apps and flows. 
Is this still recommended way to clean up connections (we have a lot of connections with Error status)?

**Comment by Jenefer-Monroe** [comment]
Yes deleting connections is a little dangerous, that is why we put it into your hands to do manually and one by one, and removed our automated processes. 

The issue in that bug you point to was around the "logic flows" connections which started incorrectly reporting themselves as error'd when they were not. the product team removed the ability to delete these connections as a result. So that risk is gone. 

Generally speaking: 
If someone has a broken connection, the app/flow will not work for them anymore until it is resolved.
Deleting it means that when they go to fix it up they will have to re-create rather than "fix". Which may mean re-attaching to a connection reference for example. 
So a little more difficult but only a few clicks once people know how.

**Comment by TomekPi** [comment]
I know it's not CoE anymore, but do you know if there is a way to get info about referenced Flow or PowerApp from collected errored connections?
I can't see such props in Get-AdminPowerAppConnection results, other cmdlets also doesn't look promising, and docs are not helpful.

**Comment by mohamrizwa** [comment]
This issue is being closed as its scope is now included in a newly created consolidated enhancement request for centralized management of orphaned components (Apps, Flows, Connection References) in the CoE Starter Kit.
New Issue Reference: https://github.com/microsoft/coe-starter-kit/issues/10331
Thank you for your contributions
